storybook/figmaio.cpp

```diff
@@ -0,0 +1,49 @@
+#include "figmaio.h"
+
+#include <QFile>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+
+QMap<QString, QList<QString>> FigmaIO::read(const QString &file)
+{
+    QFile f(file);
+    if (!f.open(QIODevice::ReadOnly | QIODevice::Text)) {
+        qWarning("FigmaIO::read - failed to open file (%s)",
+                 qUtf8Printable(file));
+        return {};
+    }
+
+    QJsonParseError error;
+    auto jsonDoc = QJsonDocument::fromJson(f.readAll(), &error);
+
+    if (error.error != QJsonParseError::NoError) {
+        qWarning("FigmaIO::read - error parsing json file (%s): %s",
+                 qUtf8Printable(file), qUtf8Printable(error.errorString()));
+        return {};
+    }
+
+    QMap<QString, QList<QString>> mapping;
+
+    if (jsonDoc.isObject()) {
+        auto rootObject = jsonDoc.object();
+
+        auto i = rootObject.constBegin();
+        while (i != rootObject.constEnd()) {
+            QJsonValue val = i.value();
+            QJsonArray links = val.toArray();
+
+            QList<QString> linksList;
+            linksList.reserve(links.size());
+
+            for (auto link : links)
```

Comment:
I love those qt subtleties ;). Funny thing: once during interview the interviewer claimed that `for (const auto link : links)` is enough to prevent detaching and I had to argue that's not true :laughing:

Reply:
Correct, that line _would_ actually detach (because of the missing constref) :)